### ISSUE TITLE
Is it possible to have nested structs and provide storage for them?

### ISSUE THREAD

**Issue by juandent** [fnc12/sqlite_orm#184]
Sometimes the need or convenience arises, to separate a group of fields into its own struct, inside another struct. I have not tried it, first I wanted to ask you if this is possible or if it complicates things more than it clears them.

Regards,
Juan

**Comment by fnc12** [comment]
I will ask a counter question: how would you store substructs in pure database?

**Comment by juandent** [comment]
Well,in a database the nested structs would need to be flattened, so it does not help there. But, in C++ code, it might make more cohesive responsbilites (Single Responsibility Principle)..

Of course the classes could just be defined outside (not nested)and then a field from that class could be embedded in the "outer" class like so:

```
class OneResponsibility
{
};

class AnotherResponsibility
{
     OneResponsibility m_one;
};
```

But then we would need to define how OneResponsability is stored (would that be like the example of the enum binding?)

The point I am trying to avoid is to have all data structures be flat (i.e. only primitive members)...

**Comment by fnc12** [comment]
1) there is nothing bad in bad structures cause it makes your code look similar with database - it bring more clarity
2) there is not built-in way to map nested structures. All you can do is to map them flattened and then create you own class derived from your storage and define your custom `get`/`get_all` etc functions and use them. I've been thinking about some kind of mapping just like it is done in [RealsSwift](https://github.com/realm/realm-cocoa) but `sqlite_orm`'s goal is to make all clear and available to tune up. Let's assume I want to add nested structures. How would developers specify subtable's name? And how to specify type of relation (one to many, many to many or just simple without foreign key). `sqlite_orm` has a great API to specify table names and column names explicitly. This is a very important thing cause most of ORM libs I met don't do it. And it becomes hard to navigate through database with a different client (CLI or any GUI app) cause you even don't know your database's tables` and columns` names. And that sucks. So if you got an idea how to make subentities with a beautiful and efficient API just as it is now please tell me. Thanks

**Comment by fnc12** [comment]
I used this GUI several times https://sqlitebrowser.org. Also I debug SQLite with built-in `sqlite3` CLI.
I was thinking about interface like this:
```c++
class Mark {
public:
    int value;
    int student_id;
};

class Student{
public:
    int id;
    std::string name;
    int roll_number;
    std::vector<decltype(Mark::value)> marks;
};

using namespace sqlite_orm;
auto storage = make_storage("subentities.sqlite",
                            make_table("students",
                                       make_column("id",
                                                   &Student::id,
                                                   primary_key()),
                                       make_column("name",
                                                   &Student::name),
                                       make_column("roll_no",
                                                   &Student::roll_number),
                                       make_many_to_one<Mark>(&Mark::student_id)),
                            make_table("marks",
                                       make_column("mark",
                                                   &Mark::value),
                                       make_column("student_id",
                                                   &Mark::student_id)));
```
This is just a draft

**Comment by juandent** [comment]
Hi,
Thank you for your recommendation though I now realize I was unclear when I mentioned user interface. What I meant was what class library do you use for data presentation and update (sort of like Microsoft's MFC or FLTK) and if you could be so kind as to send a snippet showing how you put the data model together in the View (e.g. in datagrids)...
A small representative example will push me in the right direction to be sure. 

In other words: you have provided the DB access, how do you link this with the user interface of applications? 

Deeply thankful
Juan

**Comment by fnc12** [comment]
@juandent same way how you draw strings, numbers and other visual things - just obtain data with `sqlite_orm` and adapt it to your view

**Comment by juandent** [comment]
Do you work in Windows?
Do you use MFC?

**Comment by juandent** [comment]
What I would really need is a sampleof how to make the interaction between the data model and the view... a bit more in detail please...

**Comment by fnc12** [comment]
what's your target platform and gui framework?

**Comment by juandent** [comment]
Windows but open as to the gui framework

**Comment by juandent** [comment]
can be MFC or FLTK for instance .... rather: any suggestions on your part?

**Comment by juandent** [comment]
what do you use,if I may?

**Comment by juandent** [comment]
for instance, when displaying rows in a datagrid, do you use vector<tuple<...>>?

**Comment by fnc12** [comment]
I don't use MFC and I seldom work with Windows. Most popular my platforms are mobile ones. For windows you can you any C++ GUI framework with `sqlite_orm` easily: [qt](https://www.qt.io) - heavy crossplatform lib with custom UI, [gtkmm](https://www.gtkmm.org/en/) - not so heavy but poorly documented crossplatform GUI lib. MFC and FLTK never used. I have experience in C# and .NET native windows controls. If you wanna use `sqlite_orm` inside a project with a different language (C#, Java, Swift etc) you need to make bindings to it. Also there is a good open source crossplatform C++ game framework [cocos2d-x](http://www.cocos2d-x.org) but it is not useful for developing business apps. Also one can use C++ code within objective-c classes inside Cocoa AppKit projects for macOS. Try it - Xcode is great (or it was a couple years ago).

**Comment by juandent** [comment]
I have tried XCode and it's fine but I prefer Visual Studio in Windows.
Perhaps to narrow down my question: how do you interface the result of a select with a datagrid?

Sorry for taking your time (I am stuck on this design issue)

Juan

**Comment by fnc12** [comment]
it's ok. Datagrid depends on your data and your client machine performance. If data amount is not too large and machine has a nice performance so you can just store all data in ram in some kind of vector of vectors and leave it that way. But if rows number is large and machine is weak (e.g. cell phone) so you need a cute way of displaying table data. There are some ways to optimize table views to show only visible rows and not to even store other data. E.g. [NSTableView](http://www.knowstack.com/swift-3-1-nstableview-complete-guide/)

**Comment by juandent** [comment]
Thanks @fnc12 ! Will try my hand at this... Many decisions to be made.  Great help!!
